synapse/storage/data_stores/main/end_to_end_keys.py

```diff
@@ -330,6 +332,105 @@ def get_e2e_cross_signing_key(self, user_id, key_type, from_user_id=None):
+            "  FROM e2e_cross_signing_keys k"
+            "  JOIN (SELECT user_id, MAX(stream_id) AS stream_id"
+            "          FROM e2e_cross_signing_keys"
+            "         WHERE keytype = ?"
+            "         GROUP BY user_id) s"
+            " USING (user_id, stream_id)"
+            " WHERE k.user_id IN (%s) AND k.keytype = ?"
+        ) % (",".join("?" for u in users))
+        query_params = [key_type]
+        query_params.extend(users)
+        query_params.append(key_type)
+
+        txn.execute(sql, query_params)
+        rows = self.cursor_to_dict(txn)
+
+        result = {}
+        devices = {}
+        for row in rows:
+            user_id = row["user_id"]
+            key = json.loads(row["keydata"])
+            result[user_id] = key
+            for k in key["keys"].values():
+                devices[user_id] = k
+
+        if devices and from_user_id:
+            # if we're asked to get signatures, and we have any devices to get
+            # signatures for, fetch the signatures
+            sql = (
+                "SELECT target_user_id, key_id, signature "
+                "  FROM e2e_cross_signing_signatures "
+                " WHERE user_id = ?"
+                "   AND (%s)"
+            ) % (
+                " OR ".join(
+                    "(target_user_id = ? AND target_device_id = ?)" for d in devices
+                )
+            )
+            query_params = [from_user_id]
+            for item in devices.items():
+                # item is a (user_id, device_id) tuple
+                query_params.extend(item)
+
+            txn.execute(sql, query_params)
+            rows = self.cursor_to_dict(txn)
+
+            # and add the signatures to the appropriate keys
+            for row in rows:
+                key_id = row["key_id"]
+                target_user_id = row["target_user_id"]
+                target_user_key = result[target_user_id]
+                signatures = target_user_key.setdefault("signatures", {})
+                user_sigs = signatures.setdefault(from_user_id, {})
+                user_sigs[key_id] = row["signature"]
+
+        return result
+
+    def get_e2e_cross_signing_keys_bulk(self, users, key_type, from_user_id=None):
+        """Returns the cross-signing keys for a set of users.
+
+        Args:
+            users (iterable[str]): the users whose keys are being requested
+            key_type (str): the type of keys that are being requested: either 'master'
+                for a master key, 'self_signing' for a self-signing key, or
+                'user_signing' for a user-signing key
+            from_user_id (str): if specified, signatures made by this user on
+                the self-signing keys will be included in the result
+
+        Returns:
+            Deferred[dict] map of user ID to key data.  If a user's cross-signing key
+                was not found, their user ID will not be in the dict.
+        """
+        return self.runInteraction(
```

Comment:
You probably want to do this in batches, as if this gets called with 1000s of users then the SQL might become too big